addons/test_website/tests/test_page.py

```diff
@@ -0,0 +1,54 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo.tests import HttpCase, tagged
+from odoo.tests.common import HOST
+from odoo.tools import config
+
+
+@tagged('-at_install', 'post_install')
+class WithContext(HttpCase):
+    def test_01_homepage_url(self):
+        # Setup
+        website = self.env['website'].browse([1])
+        website.write({
+            'name': 'Test Website',
+            'domain': f'http://{HOST}:{config["http_port"]}',
+            'homepage_url': '/unexisting',
+        })
+        home_url = '/'
+        contactus_url = '/contactus'
+        contactus_url_full = website.domain + contactus_url
+        contactus_content = b'content="Contact Us | Test Website"'
+        self.env['website.menu'].search([
+            ('website_id', '=', website.id),
+            ('url', '=', contactus_url),
+        ]).sequence = 1
+
+        # 404 shouldn't be served but fallback on first menu
+        # -------------------------------------------
+        # / page exists | first menu  |  homepage_url
+        # -------------------------------------------
+        #    yes        | /contactus  |  /unexisting
+        # -------------------------------------------
+        r = self.url_open(home_url)
+        self.assertEqual(r.status_code, 200)
+        self.assertEqual(r.history[0].status_code, 303)
+        self.assertEqual(r.url, contactus_url_full)
+        self.assertIn(contactus_content, r.content)
+
+        # same with 403
+        # -------------------------------------------
+        # / page exists | first menu  |  homepage_url
+        # -------------------------------------------
+        #    yes        | /contactus  |  /test_website/200/name-1
+        # -------------------------------------------
+        rec_unpublished = self.env['test.model'].create({
+            'name': 'name',
+            'website_published': False,
+        })
+        website.homepage_url = f"/test_website/200/name-{rec_unpublished.id}"
+        r = self.url_open(home_url)
+        self.assertEqual(r.status_code, 200)
+        self.assertEqual(r.history[0].status_code, 303)
+        self.assertEqual(r.url, contactus_url_full)
+        self.assertIn(contactus_content, r.content)
```

Comment:
Not sure to fully understand :p 

Yes it tests that both a 403 or 404 can't be served and will fallback on the first menu.

And both cases need a specific test. You can see it by removing one of the exception in main.py (below in this PR)
```py
try:
    return request._serve_ir_http()
except (AccessError, NotFound):
    pass
```

If you remove AccessError and the test case you commented on, it won't fail

Maybe you meant that the test is not ensuring the `/test_website/200/name-1` is really a 403? I will add an extra asset for that :+1:

Reply:
> Maybe you meant that the test is not ensuring the /test_website/200/name-1 is really a 403? I will add an extra asset for that +1

That's what I meant yes. It's the same for 404 indeed but relying on the fact /notexisting leads to a 404 seems more straightforward/robust enough. Relying on /test_website/200/name-1 leading to a 403 seemed more fragile but maybe that's just because I did not know of our /test_website/ routes before reading the test.

Reply:
Oh, it is because the record is unpublished so it becomes unreachable/403.
That's the default value (unpublished) but I also added explicitly `'website_published': False,` when creating the record + named the variable `rec_unpublished`, maybe it is not straightforward to figure still

But I added an extra step to ensure the 403 and 404 are reached, for both cases:
```py
r = self.url_open(website.homepage_url)
self.assertEqual(r.status_code, 404, "The website homepage_url should be a 404")
```